### PR TITLE
Diffusion-models

### PR DESCRIPTION
**Pull request by MBrede**
Wrote the chapter.
Task-notebooks and an illustration will be created tomorrow.

### CODE REVIEW

File: content/diff_models.qmd [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+The authors also show that this method can be used to transfer the learned representations to other tasks, such as zero-shot classification.
+
+## Diffusion Models
+
+Though models like DALL-E and CLIP represented significant milestones in the journey of text-to-image generation, the field continued to evolve rapidly, leading to the advent of Stable Diffusion. This evolution was partly inspired by the need for more control over the generation process and a desire for higher-quality outputs at lower computational costs.
+
+The GAN-architecture (first published in 2014) was the de-facto standard for quite some time and though the central principle of their successors diffusion models was published in 2015 [@sohl-dicksteinDeepUnsupervisedLearning2015], it took until 2020 for them to beat GANs on most benchmarks [@dhariwalDiffusionModelsBeat2021].
+
+The diffusion model's central principle is the training on a sequence of gradually noised images. Here's the updated snippet with the two reasons for step-by-step noise addition included:
```
Comment (alwinklick) [inline content/diff_models.qmd:47]: almost word-for-word repetition.

The diffusion model's central principle is the training on a sequence of gradually noised images. Here's the updated snippet with the two reasons for step-by-step noise addition included:

The diffusion model's central principle is training on a sequence of gradually noised images.

File: content/diff_models.qmd [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+This architecture has been widely adopted and is used as a foundation^[Or at least as an orientation.] for many state-of-the-art text-to-image models, including Stable Diffusion, as well as DALL-E 2.
+
+::: {.callout-note}
+
+## 📝 Task 
+
+Test out a SD-model!
+
+Use the colab-Notebook you can find [here]() to test out Stable Diffusion using the huggingface `diffusers`-module and generate some images.
```
Comment (alwinklick) [inline content/diff_models.qmd:86]: missing link?

File: content/diff_models.qmd [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+The paper, which introduced a contrastive^[Contrastive also being the namesake of the method (Contrastive Language-Image Pre-training)] method to learn visual representations from images and text descriptions, bridged the gap between image and text embeddings. This contrastive principle is illustrated in @fig-CLIP.
+
+![Illustration of the contrastive learning paradigm used in CLIP, taken from @radfordLearningTransferableVisual2021](../imgs/clip.png){#fig-CLIP .enlarge-onhover}
+
+A matrix of all combinations of images and text descriptions is created. The model then learns to predict the correct image for a given text description and vice versa. This is done by encoding both the image and the text into a vector space, which is then used to calculate the similarity between the two vectors.
+to do this, both a vision- and a text-transformer are trained as encoders to maximize the cosine similarity between the encoded image and text for each pair in the matrix and minimizing it for all other pairs.
+The authors also show that this method can be used to transfer the learned representations to other tasks, such as zero-shot classification.
+
+## Diffusion Models
```
Comment (alwinklick) [inline content/diff_models.qmd:41]: very nice! i learned a lot new stuff!

File: content/diff_models.qmd [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+4. Test the `num_inference_steps`-parameter. How does it affect the quality of the generated image?
+
+:::
+
+## Multimodal Models
+
+So called multimodal models are models that are trained to fit one latent distribution for multiple modalities. This means that instead of only using the image encoder and decoder with some kind of informed diffusion model to generate images in between, encoders and decoders for multiple modalities are trained to map onto the same latent space. This results in a family of models that can take inputs in multiple modalities and create outputs in a similar fashion.
+There are different approaches to solve this task, of which two will be discussed in the following section
+
+### Unidiffuser
```
Comment (alwinklick) [inline content/diff_models.qmd:103]: excellent!

File: content/diff_models.qmd [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+For image-decoding, the Stable Diffusion VAE-decoder is used to map the latent space back into an image. For text-decoding, a GPT-2-based [@radfordLanguageModelsAre2019]model is finetuned to take the latent space embeddings as a prefix-embedding and to autoregressively generate text.
+During finetuning, the CLIP-embeddings were held constant and only the GPT-2-parameters were updated. This means that the already defined latent space learned by the CLIP-model is used to map the GPT-2 decoder onto it.
+
+These embeddings are then used to train a U-ViT [@baoAllAreWorth2022] model, which takes the concated time-step-tokens, noised text- and image-embeddings as input-tokens and outputs the estimated noise-vector for the denoising process.
+
+::: {.callout-note}
+
+## 📝 Task 
+
+Use [this]() colab notebook to test out Unidiffuser using the huggingface `diffusers`-module and generate some images and text.
```
Comment (alwinklick) [inline content/diff_models.qmd:121]: also link

File: content/diff_models.qmd [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+Try the tips you tested on the basic SD-model and test whether the model accurately generates descriptions for you generated images.
+
+
+Present your results of both tasks to the course and upload your adapted notebook to moodle.
+:::
+
+### Llama 3.2
```
Comment (alwinklick) [inline content/diff_models.qmd:129]: very good!

File: content/diff_models.qmd
Comment (alwinklick) [inline content/diff_models.qmd:1]: Very good addition, i learned a lot of new stuff!

**Review by alwinklick — APPROVED** [review]
A very good chapter indeed! 

found only minor issues, feel free to fix to make it even more awesomer!

File: content/diff_models.qmd [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+GANs are the architecture behind the page [thispersondoesnotexist.com](www.thispersondoesnotexist.com) and its [clones](https://thisxdoesnotexist.com/).
+
+### DALL-E
+
+The raining position of GANs as the de-facto standard for AI image generation was challenged by the release of DALL-E by OpenAI in January 2021.
+DALL-E is a text-to-image model, which means that it can generate images based on a text description.
+
+This model was trained on a dataset containing image-caption pairs in two parts:
+
+1. A Variational Autoencoder (VAE)^[Since the latent space these images are compressed to is of a defined set of classes, the authors call the model a discrete VAE which makes a lot of sense.] to compress the image data into a latent space. This means, that each image was compressed into a 32x32 grid, for which each grid cell was encoded as a discrete probability distribution with 8192 dimensions. This latent "token"-space is, although the architecture is pretty different, quite close to what our text-transformers outputted in the MLM-task.
```
[the remark this answers is not in the document]
Comment (MBrede) [reply content/diff_models.qmd:22]: To have a mapping to the token-space, as far as I understood. The variable has to be discretized anyway and this latent space has the VAE-magic as far as I understood it.

File: content/diff_models.qmd [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+2. A Transformer to learn the relationship between text-captions and the latent space. This was done by encoding images using the pretrained VAE und argmax choosing the 32x32-token-representation of the image. The text-captions were limited to 256 tokens and concatenated with the 1024-dimensional image-tokens. The model is then trained to predict the next token in the sequence, which is either a text or an image token, similarly to the learning-paradigm we discussed when talking about the transformer-training.
+
+The resulting 1024 image-tokens can then be fed into the decoder-Block of the VAE to generate an image.
+An illustration of the training-process can be seen in @fig-dVAE.
+
+<!-- ![Illustration of the DALL-E training process.](../imgs/dalle_architecture.png){#fig-dVAE .enlarge-onhover} -->
```
[the remark this answers is not in the document]
Comment (MBrede) [reply content/diff_models.qmd:28]: Ill add it just now

File: content/diff_models.qmd (hunk 1 above)
Comment (MBrede) [reply content/diff_models.qmd:47]: however did that happen :thinking_face:

File: content/diff_models.qmd (hunk 2 above)
Comment (MBrede) [reply content/diff_models.qmd:86]: Ill add it just now

File: _quarto.yml [hunk 9]
```diff
@@ -33,8 +33,9 @@ book:
         - content/agent_interaction.qmd
     - part: "Image Generation"
       chapters:
-        - content/generator_basics.qmd
-        - content/augmentation.qmd
+        - content/diff_models.qmd
+        - content/gans_and_augmentation.qmd
+        - content/generation_in_agent_pipelines.qmd
```
[the remark this answers is not in the document]
Comment (MBrede) [reply _quarto.yml:38]: exactly

**Pull request merged**